src/js/bs3/module/HintPopover.js

```diff
@@ -0,0 +1,273 @@
+      var content = this.content($activeItem.data('index'), $activeItem.data('item'));
+
+      if (typeof content === 'string') {
+        content = document.createTextNode(content);
+      }
+
+      $popover.removeData('wordRange');
+
+      wordRange.insertNode(content);
+      range.createFromNode(content).collapse().select();
+    };
+
+    this.content = function (index, obj) {
+      if (hint[index] && hint[index].content) {
+        return hint[index].content(obj);
+      } else {
+        return obj;
+      }
+    };
+
+    this.searchKeyword = function (hint, keyword, callback) {
+      if (hint && hint.match.test(keyword)) {
+        var matches = hint.match.exec(keyword);
+        this.search(hint, matches[1], callback);
+      } else {
+        callback();
+      }
+    };
+
+    this.search = function (hint, keyword, callback) {
+
+      if (hint && hint.search) {
+        hint.search(keyword, callback);
+      } else {
+        callback();
+      }
+
+    };
+
+    this.createListTemplate = function (index, hint, list) {
+      var items  = [];
+      list = list || [];
+
+      for (var i = 0, len = list.length; i < len; i++) {
+        var $item = $('<div class="hint-item"></div>');
+        $item.append(this.createItemTemplate(hint, list[i]));
+        $item.data({ 'index' : index, 'item' : list[i] });
+        items.push($item);
+      }
+
+      if (index === 0 && items.length) {
+        items[0].addClass('active');
+      }
+
+      return items;
+    };
+
+    this.createItemTemplate = function (hint, obj) {
+      if (hint && hint.template) {
+        return hint.template(obj);
+      } else {
+        return '';
+      }
+    };
+
+    this.updateKeydown = function (e) {
+      if ($popover.css('display') === 'none') {
+        return true;
+      }
+
+      if (e.keyCode === KEY.ENTER) {
+        e.preventDefault();
+        // replace
+        this.replace();
+        this.hide();
+        summernote.invoke('editor.focus');
+      } else if (e.keyCode === KEY.UP) {
+        e.preventDefault();
+        // popover cursor up
+        this.moveUp();
+      } else if (e.keyCode === KEY.DOWN) {
+        e.preventDefault();
+        // popover cursor down
+        this.moveDown();
+      }
+    };
+
+    this.invokeHint = function (index, hint, keyword) {
+      this.searchKeyword(hint, keyword, function (list) {
+        var $group = $popoverContent.find('.hint-group-' + index);
+        var templateList = self.createListTemplate(index, hint, list);
+        $group.html(templateList);
+
+        if (templateList.length) {
+          self.show();
+        }
+      });
+    };
+
+    this.update = function (e) {
+      if (!hint) {
+        return false;
+      }
+
+      if (DROPDOWN_KEYCODES.indexOf(e.keyCode) > -1) {
+        if (e.keyCode === KEY.ENTER) {
+          if ($popover.css('display') === 'block') {
+            return false;
+          }
+        }
+
+      } else {
+
+        if (hint && hint.length) {
```

Comment:
hint is run by hint option. 

refer to examples/hint-emoji.html file.

so hint can't run code if hint options is not exists.